sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/XmlExternalEntityShouldNotBeParsed.cs

```diff
@@ -0,0 +1,187 @@
+                symbol.ContainingType.GetSymbolType().IsAny(UnsafeXmlResolvers);
+        }
+
+        [SuppressMessage("AnalyzerCorrectness", "RS1001:Missing diagnostic analyzer attribute.", Justification = "This is not a different diagnostic.")]
+        private class XmlTextReaderShouldBeSafe : ObjectShouldBeInitializedCorrectlyBase
+        {
+            private NetFrameworkVersion NetFrameworkVersion;
+
+            private INetFrameworkVersionProvider VersionProvider;
+
+            public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(rule);
+
+            private static readonly ISet<string> TrackedProperties = new HashSet<string>
+            {
+                "XmlResolver", // should be null
+                "DtdProcessing", // should not be Parse
+                "ProhibitDtd" // should be true in .NET 3.5
+            }.ToImmutableHashSet();
+
+            protected override bool IsTrackedPropertyName(string propertyName) => TrackedProperties.Contains(propertyName);
+
+            internal override ImmutableArray<KnownType> TrackedTypes { get; } = ImmutableArray.Create(KnownType.System_Xml_XmlTextReader);
+
+            // FIXME create constructor that gets the target framework and IsAllowedValue should consider that
+            protected override bool CtorInitializesTrackedPropertyWithAllowedValue(ArgumentListSyntax argumentList, SemanticModel semanticModel) =>
+                NetFrameworkVersion switch
+                {
+                    NetFrameworkVersion.After45 => true,
+                    NetFrameworkVersion.Between4And45 => false,
+                    NetFrameworkVersion.Below4 => false,
+                    _ => true
+                };
+
+            // FIXME create constructor that gets the target framework and IsAllowedValue should consider that
+            protected override bool IsAllowedValue(ISymbol symbol) =>
+                !IsUnsafeXmlResolverConstructor(symbol) &&
+                !symbol.GetSymbolType().IsAny(UnsafeXmlResolvers);
+
+            protected override bool IsAllowedValue(object constantValue)
+            {
+                if (constantValue == null)
+                {
+                    return true;
+                }
+                if (constantValue is int integerValue)
+                {
+                    return integerValue != DtdProcessingParse;
+                }
+                return constantValue is bool && (bool)constantValue;
+            }
+
+            protected override void CompilationAction(Compilation compilation)
+            {
+                NetFrameworkVersion = VersionProvider.GetDotNetFrameworkVersion(compilation);
+            }
+
+            internal XmlTextReaderShouldBeSafe(INetFrameworkVersionProvider netFrameworkVersionProvider)
+            {
+                VersionProvider = netFrameworkVersionProvider;
+            }
+
+            private bool IsUnsafeXmlResolverConstructor(ISymbol symbol) =>
```

Comment:
This method duplicates the code from **XmlDocumentShouldBeSafe**. We can extract it to `XmlExternalEntityShouldNotBeParsed` to be able to reuse it.